classes/helpers/product-operation.php

```diff
@@ -0,0 +1,329 @@
+			else
+				return Shop::getShops(true);
+		}
+		else
+		{
+			$ctx = Context::getContext();
+			return array(
+				(int)$ctx->shop->id => array(
+					'id_shop' => (int)$ctx->shop->id,
+					'id_shop_group' => (int)$ctx->shop->id_shop_group,
+				),
+			);
+		}
+	}
+
+	/**
+	 * Loads a Nosto product model for given PS product ID, language ID and shop ID.
+	 *
+	 * @param int $id_product the PS product ID.
+	 * @param int $id_lang the language ID.
+	 * @param int $id_shop the shop ID.
+	 * @return NostoTaggingProduct|null the product or null if could not be loaded.
+	 */
+	protected function loadNostoProduct($id_product, $id_lang, $id_shop)
+	{
+		$product = new Product($id_product, false, $id_lang, $id_shop);
+		if (!Validate::isLoadedObject($product))
+			return null;
+
+		if (isset($product->visibility) && $product->visibility === 'none')
+			return null;
+
+		$this->makeContextSnapshot();
+
+		$nosto_product = new NostoTaggingProduct();
+		$nosto_product->loadData($this->makeContext($id_lang, $id_shop), $product);
+
+		$this->restoreContextSnapshot();
+
+		$validator = new NostoValidator($nosto_product);
+		if (!$validator->validate())
+			return null;
+
+		return $nosto_product;
+	}
+
+	/**
+	 * Stores a snapshot of the current context.
+	 */
+	protected function makeContextSnapshot()
+	{
+		$this->_context_snapshot = array(
+			'shop_context' => (_PS_VERSION_ >= '1.5') ? Shop::getContext() : null,
+			'context_object' => Context::getContext()->cloneContext()
+		);
+	}
+
+	/**
+	 * Restore the context snapshot to the current context.
+	 */
+	protected function restoreContextSnapshot()
+	{
+		if (!empty($this->_context_snapshot))
+		{
+			$original_context = $this->_context_snapshot['context_object'];
+			$shop_context = $this->_context_snapshot['shop_context'];
+			$this->_context_snapshot = null;
+
+			$current_context = Context::getContext();
+			$current_context->language = $original_context->language;
+			$current_context->shop = $original_context->shop;
+			$current_context->link = $original_context->link;
+			$current_context->currency = $original_context->currency;
+
+			if (_PS_VERSION_ >= '1.5')
+			{
+				Shop::setContext($shop_context, $current_context->shop->id);
+				Dispatcher::$instance = null;
+				if (method_exists('ShopUrl', 'resetMainDomainCache'))
+					ShopUrl::resetMainDomainCache();
+			}
+		}
+	}
+
+	/**
+	 * Modifies the current context and replaces the info related to shop, link, language and currency.
+	 *
+	 * We need this when generating the product data for the different shops and languages.
+	 * The currency will be the first found for the shop, but it defaults to the PS default currency
+	 * if no shop specific one is found.
+	 *
+	 * @param int $id_lang the language ID to add to the new context.
+	 * @param int $id_shop the shop ID to add to the new context.
+	 * @return Context the new context.
+	 */
+	protected function makeContext($id_lang, $id_shop)
+	{
+		if (_PS_VERSION_ >= '1.5')
```

Comment:
This method is a bit hacky to get the product page url and image url to have the correct domain, path and "friendly url" settings. We run this per shop, so we need to replace the context for each shop.